src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,477 @@
+    double* params) const {
+  THROW_CHECK(HasValidFullCovariance());
+  THROW_CHECK(HasBlock(params));
+  int index = GetBlockIndex(params);
+  int num_params = GetBlockTangentSize(params);
+  return cov_variables_.block(index, index, num_params, num_params);
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    const std::vector<double*>& blocks) const {
+  THROW_CHECK(HasValidFullCovariance());
+  std::vector<int> indices;
+  for (const auto& block : blocks) {
+    THROW_CHECK(HasBlock(block));
+    int index = GetBlockIndex(block);
+    int num_params_pose = GetBlockTangentSize(block);
+    for (int i = 0; i < num_params_pose; ++i) {
+      indices.push_back(index + i);
+    }
+  }
+#if EIGEN_VERSION_AT_LEAST(3, 4, 0)
+  Eigen::MatrixXd output = cov_variables_(indices, indices);
+#else
+  size_t n_indices = indices.size();
+  Eigen::MatrixXd output(n_indices, n_indices);
+  for (size_t i = 0; i < n_indices; ++i) {
+    for (size_t j = 0; j < n_indices; ++j) {
+      output(i, j) = cov_variables_(indices[i], indices[j]);
+    }
+  }
+#endif
+  return output;
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    double* params1, double* params2) const {
+  THROW_CHECK(HasValidFullCovariance());
+  THROW_CHECK(HasBlock(params1));
+  THROW_CHECK(HasBlock(params2));
+  int index1 = GetBlockIndex(params1);
+  int num_params_block1 = GetBlockTangentSize(params1);
+  int index2 = GetBlockIndex(params2);
+  int num_params_block2 = GetBlockTangentSize(params2);
+  return cov_variables_.block(
+      index1, index2, num_params_block1, num_params_block2);
+}
+
+bool BundleAdjustmentCovarianceEstimatorCeresBackend::ComputeFull() {
+  ceres::Covariance::Options options;
+  ceres::Covariance covariance_computer(options);
+  std::vector<const double*> parameter_blocks;
+  for (const auto& block : pose_blocks_) {
+    parameter_blocks.push_back(block);
+  }
+  for (const auto& block : other_variables_blocks_) {
+    parameter_blocks.push_back(block);
+  }
+  try {
+    covariance_computer.Compute(parameter_blocks, problem_);
+  } catch (const std::exception& e) {
+    return false;
+  }
+  int num_params = num_params_poses_ + num_params_other_variables_;
+  Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> covs(
+      num_params, num_params);
+  covariance_computer.GetCovarianceMatrixInTangentSpace(parameter_blocks,
+                                                        covs.data());
+  cov_variables_ = covs;
+  cov_poses_ = cov_variables_.block(0, 0, num_params_poses_, num_params_poses_);
+  return true;
+}
+
+bool BundleAdjustmentCovarianceEstimatorCeresBackend::Compute() {
+  ceres::Covariance::Options options;
+  ceres::Covariance covariance_computer(options);
+  std::vector<const double*> parameter_blocks;
+  for (const auto& block : pose_blocks_) {
+    parameter_blocks.push_back(block);
+  }
+  try {
+    covariance_computer.Compute(parameter_blocks, problem_);
```

Comment:
I thought ceres doesn't use exceptions? Should this instead do `if (!covariance_computer.Compute(...)) { return false };`?

Reply:
Yes you are right. Thanks. This is updated.